include/swift/Sema/CSTrail.h

```diff
@@ -0,0 +1,169 @@
+//===--- CSTrail.h - Constraint Solver Trail --------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the \c SolverTrail class, which records the decisions taken
+// while attempting to find a solution.
+//
+//===----------------------------------------------------------------------===//
+#ifndef SWIFT_SEMA_CSTRAIL_H
+#define SWIFT_SEMA_CSTRAIL_H
+
+#include <vector>
+
+namespace llvm {
+class raw_ostream;
+}
+
+namespace swift {
+
+class TypeBase;
+class TypeVariableType;
+
+namespace constraints {
+
+class Constraint;
+
+class SolverTrail {
+public:
+
+  /// The kind of change made to the graph.
+  enum class ChangeKind {
+    /// Added a type variable to the constraint graph.
+    AddedTypeVariable,
+    /// Added a new constraint to the constraint graph.
+    AddedConstraint,
+    /// Removed an existing constraint from the constraint graph.
+    RemovedConstraint,
+    /// Extended the equivalence class of a type variable in the constraint graph.
+    ExtendedEquivalenceClass,
+    /// Added a fixed binding for a type variable in the constraint graph.
+    BoundTypeVariable,
+    /// Introduced a type variable's fixed type to inference.
+    IntroducedToInference,
+    /// Set the fixed type or parent and flags for a type variable.
+    UpdatedTypeVariable,
+  };
+
+  /// A change made to the constraint system.
+  ///
+  /// Each change can be undone (once, and in reverse order) by calling the
+  /// undo() method.
+  class Change {
+  public:
+    /// The kind of change.
+    ChangeKind Kind;
+
+    union {
+      TypeVariableType *TypeVar;
+      Constraint *TheConstraint;
+
+      struct {
+        /// The type variable whose equivalence class was extended.
+        TypeVariableType *TypeVar;
+
+        /// The previous size of the equivalence class.
+        unsigned PrevSize;
+      } EquivClass;
+
+      struct {
+        /// The type variable being bound to a fixed type.
+        TypeVariableType *TypeVar;
+
+        /// The fixed type to which the type variable was bound.
+        TypeBase *FixedType;
+      } Binding;
+
+      struct {
+        /// The type variable being updated.
+        TypeVariableType *TypeVar;
+
+        /// The representative of the equivalence class, or the fixed type.
+        llvm::PointerUnion<TypeVariableType *, TypeBase *> ParentOrFixed;
+
+        /// The saved value of TypeVariableType::Implementation::getRawOptions().
+        unsigned Options;
+      } Update;
+    };
+
+    Change() : Kind(ChangeKind::AddedTypeVariable), TypeVar(nullptr) { }
```

Comment:
Do we still need this?

Reply:
It's needed because eg this calls the default constructor and it won't synthesize one for us because of the unions:

```
SolverTrail::Change
SolverTrail::Change::addedConstraint(Constraint *constraint) {
  Change result;
  result.Kind = ChangeKind::AddedConstraint;
  result.TheConstraint = constraint;
  return result;
}
```

If there's another way to initialize this without the default constructor I can refactor it later, because I'll be adding lots more Change kinds soon. :)